### ISSUE TITLE
False alarm when using "return null" and type is known to support nulls

### ISSUE THREAD

**Issue by vlsi** [typetools/checker-framework#3614]
Sample code:

```java
  public static @PolyNull Boolean not(@PolyNull Boolean b) {
    return (b == null) ? null : !b;
  }
```

Error:

```
SqlFunctions.java:2571: error: [return.type.incompatible] incompatible types in return.
    return (b == null) ? null : !b;
                       ^
  type of expression: @Initialized @Nullable Boolean
  method return type: @Initialized @PolyNull Boolean
```

Could checker be a bit smarter, so it could see that {{b == null}} implies {{@PolyNull}} is in {{@Nullable}} mode, so it should be perfectly valid to return {{null}}.

I understand I can rewrite the method like {{return (b == null) ? b : !b;}}, however, it would be nice if checker could figure out that automatically.

Another example might involve regular generics without Poly annotations:

```java
  public <T> static T of(T a) {
    return a == null ? null : a;
  }
```

It would be nice if checker could figure out `return null` is reachable only in the case `T is nullable`.

Another example:

```java
  /** Converts a JDBC array to a list. */
  public static @PolyNull List arrayToList(final java.sql.@PolyNull Array a) {
    if (a == null) {
      return null; // <- error here
    }
    ...
```

**Comment by vlsi** [comment]
One more sample which does not have a nice workaround except `castNonNull`

```java
  public static @PolyNull Integer plus(@PolyNull Integer b0, @PolyNull Integer b1) {
    return (b0 == null || b1 == null) ? null : (b0 + b1);
  }
```

**Comment by mernst** [comment]
The `not` and `plus` examples may be related to #3622, because rewriting them as `if` statements causes the Nullness Checker to accept them.

**Comment by mernst** [comment]
Followup: the fix for #3622 does not resolve this issue.
There are test cases checked into the repository in file `Issue3614.java`.
